benchmark_test/regression.go

```diff
@@ -0,0 +1,56 @@
+package benchmark
+
+import (
+	"fmt"
+
+	"github.com/elastic/hey-apm/errs"
+	"github.com/elastic/hey-apm/es"
+	"github.com/elastic/hey-apm/models"
+
+	"github.com/elastic/hey-apm/conv"
+	"github.com/elastic/hey-apm/types"
+	"github.com/pkg/errors"
+)
+
+// compares the given report with saved reports with the same input indexed in the last specified days
+// returns an error if connection can't be established,
+// or performance decreased by a margin larger than specified
+func VerifyNoRegression(conn es.Connection, r models.Report, margin float64, days string) error {
+	inputMap := conv.ToMap(r.Input)
+	filters := []types.M{
+		{
+			"range": types.M{
+				"@timestamp": types.M{
+					"gte": fmt.Sprintf("now-%sd/d", days),
+					"lt":  "now",
+				},
+			},
+		},
+	}
+	for k, v := range inputMap {
+		filters = append(filters, types.M{"match": types.M{k: v}})
+	}
+	body := types.M{
+		"query": types.M{
+			"bool": types.M{
+				"must": filters,
+			},
+		},
+	}
+
+	savedReports, fetchErr := es.FetchReports(conn, body)
+
+	var regression error
+	for _, sr := range savedReports {
+		if r.ReportId != sr.ReportId && r.Performance()*margin < sr.Performance() {
```

Comment:
I haven't given this a lot of thoughts yet, but that basically means that if each of n changes over the last couple of days has decreased performance a bit, all of them are fine. But at some point one change is not within the margin boundary compared to the oldest fetched report, and then a regression is raised, although this specific change might not be the general problem.  
I was wondering if you've looked into alternatives or how the agent devs check this?

Reply:
> although this specific change might not be the general problem

correct, but you would look into it, inspect dashboards, commit history, etc. the important thing is that we get notified when performance decreases bellow a margin.

as far as i know, agent devs don't check for regressions at all (i might be wrong)

Reply:
Fair enough. I was simply curious how others do it and if you have investigated other options, especially having in mind that we might want to be able to trigger this for specific PRs at some point.

Reply:
Right now we have _nothing_, so if we get _something_, it is going to be better imo. 
If we find out that it is not good enough, we can iterate. 
Adjusting the number of days to look back and the error margin should also give some degree of flexibility.

Reply:
Ofc that's all true, and it is a great start. I was only curious since I haven't looked into that at all.